open_lm/file_utils.py

```diff
@@ -93,33 +94,88 @@ def get_metadata_file(path):
     out = [json.loads(o) for o in out.decode('utf-8').split('\n')[:-1]]
     return out
 
-def get_shards_for_epoch(num_samples, epoch, path):
+def get_shards_for_chunk(num_samples, chunk, path):
+    """ Function to get a chunk of shards to train on.
+
+    Chunks are groups of shards with samples roughly equal to the number of samples
+    that will be seen during training. This function uses the dataset manifest
+    to split the shards into chunks, and assign shards to each chunk.
+    """
     metadata = get_metadata_file(path)
     shard_list = []
     curr_shard_list = []
     chunk_count_list = []
     curr_chunk_count = 0
-    real_chunk_count = 0
     for m in metadata:
         curr_chunk_count += m['num_chunks']
-        real_chunk_count += m['num_chunks']
         curr_shard_list.append(m['shard'])
         if curr_chunk_count >= num_samples:
             shard_list.append(curr_shard_list)
-            curr_chunk_count = curr_chunk_count - num_samples
+            chunk_count_list.append(curr_chunk_count)
             curr_shard_list = []
-            chunk_count_list.append(real_chunk_count)
-            real_chunk_count = 0
+            curr_chunk_count = 0
     
-    return shard_list[epoch % len(shard_list)], chunk_count_list[epoch % len(chunk_count_list)]
-
-def get_string_for_epoch(num_samples, epoch, path):
-    shard_list, num_samples = get_shards_for_epoch(num_samples, epoch, path)
-    shard_root = '/'.join(path.split('/')[:-1]) + '/shard_'
-    shard_string = shard_root + '{' + shard_list[0] + '..' + shard_list[-1] + '}.tar'
-    if path.startswith('s3'):
-        shard_string = f'pipe:aws s3 cp {shard_string} -'
-    return shard_string, num_samples
+    # Append remaining shards
+    if len(curr_shard_list) > 0:
+        shard_list.append(curr_shard_list)
+        chunk_count_list.append(curr_chunk_count)
+
+    return shard_list[chunk % len(shard_list)], chunk_count_list[chunk % len(chunk_count_list)]
+
+
+def enough_shards(shard_lists, min_shards_needed):
+    for sl in shard_lists:
+        if len(sl) < min_shards_needed:
+            return False
+    return True
+
+
+def enough_samples(num_samples_per_source, needed_samples_per_source):
+    for i, number in enumerate(num_samples_per_source):
+        if number < needed_samples_per_source[i]:
+            return False
+    return True
+
+
+def source_exhausted(paths, shard_list_per_source):
+    for i, source in enumerate(paths):
+        data = get_metadata_file(source)
+        if len(data) < len(shard_list_per_source[i]):
+            return True
+    return False
+
+
+def get_string_for_epoch(num_samples, starting_chunk, paths, weights, min_shards_needed):
```

Comment:
This is a pretty nontrivial function now, with a seemingly benign name. Could we add a docstring and/or possibly rename? It looks to me like the right docstring is:

```
"""
Collect shards to train on for this epoch, along with metadata.

This function looks through the list of all shards and returns a subset of shards
that result in >= num_samples samples and >= min_shards_needed shards. Note that
we cannot simply compute samples based on the number of shards, since shards can
have varying numbers of samples.
"""
```

and rename to `_get_shards_for_epoch`?